File: 486_predict-the-winner.py
from typing import List
import math

class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        def maxScores(nums, isFirst):
            if(len(nums) == 1):
                if(isFirst):
                    return nums[0]
                else:
                    return 0
            # 先取
            if(isFirst):
                return max(nums[0]+maxScores(nums[1:], False), nums[-1]+maxScores(nums[:-1], False))
            # 后取
            else:
                return max(maxScores(nums[1:], True), maxScores(nums[:-1], True))
        print(maxScores(nums, True), maxScores(nums, False))
        return maxScores(nums, True)>maxScores(nums, False)


if __name__ == "__main__":
    s = Solution()
    nums = [1, 5,2]
    print(s.PredictTheWinner(nums))

File: 486_predict-the-winner.cpp
#include "basicInclude.h"

//dp
class Solution {
public:
    // int maxScores(vector<int>& nums, bool isFirst){
    //     if(nums.size() == 1){
    //         return isFirst?nums[0]:0;
    //     }
    //     if(isFirst){
    //         return max(nums[0]+maxScores(nums[]))
    //     }
    // }
    bool PredictTheWinner(vector<int>& nums) {

    }
};

//玩家1先取，玩家2后取，计算两者的最大分数
int main()
{
    vector<int> nums = {1, 2, 3, 4, 5};
    Solution s;
    cout<<s.PredictTheWinner(nums)<<endl;
    return 0;
}